### DISCUSSION TITLE
'struct __kernel_timespec' does not contain a field named 'tv_nsec' - but it does!

### DISCUSSION THREAD

**Discussion by ttsiodras** [bpftrace/bpftrace#2205] [Q&A] [answered]
My `/usr/include/linux/time_types.h` says that `struct __kernel_timespec`...

    struct __kernel_timespec {
	    __kernel_time64_t       tv_sec;                 /* seconds */
	    long long               tv_nsec;                /* nanoseconds */
    };

...has a `tv_nsec` - but bpftrace says...

    # bpftrace -e 'tracepoint:syscalls:sys_enter_clock_nanosleep { printf("%s is sleeping for %d!\n", comm, args->rqtp->tv_nsec);}' -c ./a.out 
    stdin:1:90-109: ERROR: Struct/union of type 'struct __kernel_timespec' does not contain a field named 'tv_nsec'
    tracepoint:syscalls:sys_enter_clock_nanosleep { printf("%s is sleeping for %d!\n", comm, args->rqtp->tv_nsec);}
                                                                                             ~~~~~~~~~~~~~~~~~~~

Perhaps this is also related - I can't "see" the contents of the struct:

    # bpftrace -lv 'struct __kernel_timespec'
    # 

System specs: 

- Arch Linux
- kernel 5.15.35-1-lts
- bpftrace 0.14.1-1

**Comment by viktormalik** [comment, marked as answer]
It seems that your kernel doesn't have BTF, which is why you can't see the struct fields. The solution is to use a kernel with BTF or to include the proper header with your script:
```
# bpftrace --include 'linux/time_types.h' -e 'tracepoint...
```

**Comment by ttsiodras** [comment]
Using `bpftrace --include 'linux/time_types.h'  -e 'tracepoint...`  indeed solved the problem. Thank you!
